src/coreComponents/constitutive/fluid/DeadOilFluid.hpp

```diff
@@ -0,0 +1,714 @@
+                                             arraySlice1d< real64, 0 > const & phaseVisc ) const
+{
+  // this derivative will be discarded
+  real64 derivative = 0.0;
+
+  // 1. Hydrocarbon phases: look up in the viscosity tables
+
+  for( localIndex iph = 0; iph < m_hydrocarbonPhaseOrder.size(); ++iph )
+  {
+    // get the phase index
+    localIndex const ip = m_hydrocarbonPhaseOrder[iph];
+    // interpolate in the table to get the phase viscosity, and discard the derivative
+    m_viscosityTables[iph].compute( &pressure, phaseVisc[ip], &derivative );
+  }
+
+  // 2. Water phase: use the constant viscosity provided by the user
+
+  using PT = DeadOilFluid::PhaseType;
+  integer const ipWater = m_phaseOrder[PT::WATER];
+
+  // if water is present
+  if( ipWater >= 0 )
+  {
+    // just assign the viscosity value
+    phaseVisc[ipWater] = m_waterViscosity;
+  }
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DeadOilFluidUpdate::computeViscosities( real64 pressure,
+                                             arraySlice1d< real64 > const & phaseVisc,
+                                             arraySlice1d< real64 > const & dPhaseVisc_dPres,
+                                             arraySlice2d< real64 > const & dPhaseVisc_dGlobalCompFrac ) const
+{
+  for( double & val : dPhaseVisc_dGlobalCompFrac )
+  {
+    val = 0.0;
+  }
+
+  // 1. Hydrocarbon phases: look up in the viscosity tables
+
+  for( localIndex iph = 0; iph < m_hydrocarbonPhaseOrder.size(); ++iph )
+  {
+    // get the phase index
+    localIndex const ip = m_hydrocarbonPhaseOrder[iph];
+    // interpolate in the table to get the phase viscosity and derivatives
+    m_viscosityTables[iph].compute( &pressure, phaseVisc[ip], &(dPhaseVisc_dPres)[ip] );
+  }
+
+  // 2. Water phase: use the constant viscosity provided by the user
+
+  using PT = DeadOilFluid::PhaseType;
+  integer const ipWater = m_phaseOrder[PT::WATER];
+
+  // if water is present
+  if( ipWater >= 0 )
+  {
+    // just assign the viscosity value
+    phaseVisc[ipWater] = m_waterViscosity;
+    dPhaseVisc_dPres[ipWater] = 0.0;
+  }
+}
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DeadOilFluidUpdate::compute( real64 pressure,
+                                  real64 temperature,
+                                  arraySlice1d< real64 const, 0 > const & composition,
+                                  arraySlice1d< real64, 0 > const & phaseFrac,
+                                  arraySlice1d< real64, 0 > const & phaseDens,
+                                  arraySlice1d< real64, 0 > const & phaseMassDens,
+                                  arraySlice1d< real64, 0 > const & phaseVisc,
+                                  arraySlice2d< real64, 1 > const & phaseCompFrac,
+                                  real64 & totalDens ) const
+{
+  GEOSX_UNUSED_VAR( temperature );
+  localIndex const NC = m_componentMolarWeight.size();
+  localIndex const NP = m_phaseTypes.size();
```

Comment:
Again, this is not `constexpr` so the naming convention does not match. Use `nComponents` or `nPhases` instead.

Reply:
I agree, done.